### DISCUSSION TITLE
How to synchronize multiple actors on multiple layers on demand?

### DISCUSSION THREAD

**Discussion by ezhikov** [statelyai/xstate#1614] [Q&A]
I have a `form` model that spawns multiple `field sets` and each `field set` spawns multiple `field` actors. Right now they sync with parent machine when state changes, but this creates some hiccups in React, because state of `field` changes, `field` sends update to `field set`, its state also changes, it sends update up to the `form`. But what I really need is all this syncing only at the time of submit - determine state of `field set` and pass this state up to `form`, then submit everything based on states of `field sets` (and values of `fields` of course).

I guess, I can pass some event from `form` down and wait for responses, but I'm not sure how to collect all updates. I guess, I can count that I had all responses counting them and comparing to amount of spawned actors, but it just don't feel like right solution.

Is there any better way to handle this type of interaction?

**Comment by tomByrer** [comment]
> how to collect all updates

i'm actually working on a RadioButton Group, so I have similar concerns.  The approaches I'm about to try out:
+ not using parent/child machines, but just run machines in an array & the machines will update a global (external) state.
+ parent-children without actors & 1-deep memorization the updates to reduce the updates
+ actors with memo

Could you please post a CodeSandbox with what you have so far?

**Reply by ezhikov** [reply]
I'll try to put up a sandbox, but can't promise that it will be soon. Right now I'm thinking about two possible solutions.

First is plain approach of using iteration counter. I'll send "submit" event down to each actor (and they will send it further). Each field will respond with `update` event. Parent will count, how many responses it got and will proceed to next state when all children were synchronized. While I don't like this approach much, it seems like nice starting point.

Second thing I'm thinking about is to use RxJS, and I believe it have operator I need, but I never before used Rx, so still researching.

> not using parent/child machines

Not sure how to "just run machines in an array". In runtime I'm getting some data from server and generate arbitrary number of fieldsets and this fieldsets generate arbitrary number of fields that may be, or may not be rendered. Could you explain this approach a bit further?

> parent-children without actors

How does parent-children can be used without actors?

**Comment by f-elix** [comment]
Hi!

Are you using the sync option? If so, that might be why all your children actors are sending updates to the parent on every state change. You might want to use `sendParent()` instead to update the parent whenever you need to (https://xstate.js.org/docs/guides/actors.html#syncing-and-reading-state). 

If you only need to know the state of your children on SUBMIT, you could send an event to all your actors from the parent machine, then receive an update from them (passing data if needed). There's a good example here in the docs, in the Pure Action section: https://xstate.js.org/docs/guides/actions.html#pure-action.

Do I understand your problem correctly? 

Hope this helps.

**Reply by ezhikov** [reply]
I'm not using sync, and use `sendUpdate`. I have three levels: `A` - parent machine, `B` - arbitrary amount of children, `C` - lower level. State of `B` is important only at the very end. 

Right now it works this way: 
1. `C` changes its state
1. `C` call `sendUpdate` action
1. `B` change its state and call `sendUpdate` action.
1. On each such update react calls updates _from bottom to top_. 

I'd like it to be this way:

1. `A` issues event to children. 
2. Each `B` forwards this event to children
3. Each `C` sends update back
4. After `B` collected all updates it's determinimg its state and sends update to `A`
5. React updates doesn't matter much at this point, since user doesn't interact with app anymore

Possible solutions:

- Drop `B` completely, but calculate this state on form submit. It will reduce amount of renders, but will add lot more complexity to guards in `A` and amount of data duplicated between `A` and `C` (now all this data sits in `B`)
- Make something like `waiting for updates` state. `A` passes event down and waits until specific amount of updates were collected. Same with `B` - it transitions only when all updates collected.

What bothers me in all this is that problem is not in XState itself but in `@xstate/react` + React combo.

**Reply by ezhikov** [reply]
Apparently I can issue `xstate.update` to parent from anywhere and it will sync _all_ its children, meaning that I don't need to issue separate update event from _each_ child. Also, this means that I can just `raise`  this event _from parent itself_. All I have to do is somehow remove middle layer and it will be good.

@davidkpiano could you, please, tell, how safe it would be to send `xstate.update` from parent to itself? Maybe I'm missing something?

**Reply by davidkpiano** [reply]
All the `xstate.update` event does is notify the machine that something happened (like one of the actors emitted a new state); it is a signal to "check the actors again... they might be different".

Honestly, the best thing to do is to send explicit events from the actors. `xstate.update` is too implicit for most use-cases, leading to confusion like what you're experiencing.

**Reply by ezhikov** [reply]
Good to know that there is no inside magic to it. I don't really like the idea of using internal xstate events manually (without built-in action creators), but I already have to react to 'xstate.update' in parent machines to recalculate their state and typecheck service invocations. I think I'll roll with manual `xstate.update` raise right now to get new actors state when I really need it will continue to plan some other way of communication to avoid useless rerenders.

Thank you.
